### ISSUE TITLE
F#: Improve line-based folding

### ISSUE THREAD

**Issue by rdipardo** [ScintillaOrg/lexilla#56]
The `LineContains()` auxiliary function scans too far backward for same-styled lines. This can disrupt folding under these concurrent conditions:

- a single line in a lexical style subject to folding, e.g. `SCE_FSHARP_COMMENTLINE`
- an empty next line, followed by a group of same-styled lines

For example, a line comment group does not fold if the previous non-empty line is also a line comment [^1]:

~~~
 0 400 400   // not folded
 1 400 400
 0 400 400   // first line in comment fold
 0 400 400   // second . . .
 0 400 400   // third . . .
 1 400 400
~~~

An extra blank line is currently needed to correct this:

~~~
 0 400 400   // not folded
 1 400 400
 1 400 400
 2 400 401 + // first line in comment fold
 0 401 401 | // second . . .
 0 401 400 | // third . . .
 1 400 400
~~~

Similarly, an isolated `open` statement will be counted as the head of a nearby import list, giving the fold level an extra increment:

~~~
 0 400 400   open System
 1 400 400
 2 400 401 + module FoldingTest =
 2 401 402 +     open FSharp.Quotations
 0 402 401 |     open FSharp.Reflection
 1 401 400 |
 0 400 400       () |> ignore
 1 400 400
~~~

Here is a mostly purgative refactoring of `LineContains()` that fixes the scanning range to immediately adjacent lines:

[LexFSharp-refactor-LineContains.diff.zip](https://github.com/ScintillaOrg/lexilla/files/7921562/LexFSharp-refactor-LineContains.diff.zip)

Test files will follow once this ticket has a number.

[^1]: The failing condition is `!LineContains(..., "//", lineStartPrev, ..., SCE_FSHARP_COMMENTLINE)`, i.e., `lineStartPrev` falls within the isolated line comment: 
https://github.com/ScintillaOrg/lexilla/blob/a35a59845e793d9d37d249cf097e71fecc5f4bcd/lexers/LexFSharp.cxx#L736-L737

**Comment by rdipardo** [comment]
[LexFSharp-Test-Issue56.zip](https://github.com/ScintillaOrg/lexilla/files/7921689/LexFSharp-Test-Issue56.zip)

**Comment by nyamatongwe** [comment]
I don't want to go too far in optimizing code but to me, `FoldLexicalGroup` is repeating expressions `LineContains(styler, word, linePrev | lineNext, chAttr)`. That may be clearer and faster if each call assigned to constants `onLinePrev` and `onLineNext`.

**Comment by rdipardo** [comment]
In hindsight, I was too beholden to the comment folding logic of LexBash, the original inspiration for the patch:

https://github.com/ScintillaOrg/lexilla/blob/a35a59845e793d9d37d249cf097e71fecc5f4bcd/lexers/LexBash.cxx#L1023-L1031

The general idea works well enough, but it's prone to fold level corruption when a new comment is appended to an existing one:

![LexBash_fold_level_in_comment_group](https://user-images.githubusercontent.com/59004801/151465404-f69b144e-59cf-4da0-8371-c79b35de7236.gif)

For now, a general fold toggle smooths everything over:

![LexBash_toggle_all_folds](https://user-images.githubusercontent.com/59004801/151465561-fea2e83c-76b1-41de-9820-fcd27e929ec9.gif)

I haven't reported this yet because I wouldn't know how to fix it, apart from an explicit clamp such as LexFSharp already employs:

https://github.com/ScintillaOrg/lexilla/blob/a35a59845e793d9d37d249cf097e71fecc5f4bcd/lexers/LexFSharp.cxx#L744-L745

This kludge ought to stay in for the sake of the lexing tests, but I don't recommend adulterating LexBash in this way.

**Comment by rdipardo** [comment]
Updated my prior comment to mention that the Bash comment folding issue is currently undetectable by `TestLexers`.

**Comment by rdipardo** [comment]
> `FoldLexicalGroup` is repeating expressions `LineContains(styler, word, linePrev | lineNext, chAttr)`.

Here is a further attempt with a net difference of 9 fewer lines (over the 6 removed by the first draft):

[0003-LexFSharp-refactor-LineContains-iter-2.diff.zip](https://github.com/ScintillaOrg/lexilla/files/7967110/0003-LexFSharp-refactor-LineContains-iter-2.diff.zip)

Main features:

- within the scope of `FoldLexicalGroup`:
    + call the fore and aft line checks once each, assigning the return value to `const` booleans

- within the scope of `LineContains`:
    + shorten the range of the line scanning loop by one character width to align with EOL position

- within the scope of `::Fold`:
    + converge two EOL condition checks into a single `atEOL` boolean
    + assign the result of the `open` statement line check to a `const` boolean
    + change a condition that improves line comment folding at the top of a file; previously the first two lines had to be left-justified

**Comment by nyamatongwe** [comment]
That looks good. Committed 3 along with tests.

**Comment by nyamatongwe** [comment]
Included in 5.1.5 release.
